internal/discoverer/zookeeper.go

```diff
@@ -0,0 +1,259 @@
+package discoverer
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/api7/apisix-seed/internal/conf"
+	"github.com/api7/apisix-seed/internal/core/comm"
+	"github.com/api7/apisix-seed/internal/log"
+	"github.com/api7/apisix-seed/internal/utils"
+	"github.com/go-zookeeper/zk"
+	"golang.org/x/net/context"
+	"sort"
+	"sync"
+	"time"
+)
+
+func init() {
+	Discoveries["zookeeper"] = NewZookeeperDiscoverer
+}
+
+type ZookeeperService struct {
+	Name         string
+	BindEntries  []string
+	WatchConn    *zk.Conn
+	RootPath     string
+	WatchPath    string
+	WatchContext context.Context
+	WatchCancel  context.CancelFunc
+	WatchMutex   sync.Mutex
+}
+
+type ZookeeperDiscoverer struct {
+	zkConfig   *conf.Zookeeper
+	zkServices map[string]*ZookeeperService
+	eventMutex sync.Mutex
+
+	msgCh chan *comm.Message
+}
+
+func (zd *ZookeeperDiscoverer) Stop() {
+	for _, service := range zd.zkServices {
+		zd.unsubscribe(service)
+	}
+	close(zd.msgCh)
+}
+
+func (zd *ZookeeperDiscoverer) Query(query *comm.Query) error {
+	values, _, err := query.Decode()
+	if err != nil {
+		return err
+	}
+
+	event, entity, serviceName := values[0], values[1], values[2]
+
+	switch event {
+	case utils.EventAdd:
+		zd.eventMutex.Lock()
+		defer zd.eventMutex.Unlock()
+		err = zd.fetchService(serviceName, []string{entity})
```

Comment:
We should add the lock inside the function instead of where the function is called. This is error prone because you can't guarantee that you'll get a lock every time you call it